### PR TITLE
integration: Make SPO deployment optional

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
The SPO is required for TestAuditSeccomp(), make it optional in case
somebody wants to run tests that don't require this component.

### CODE REVIEW

**Review by eiffel-fl — COMMENTED** [review]
It seems cool but I have one concern:

File: integration/integration_test.go [hunk 1]
```diff
@@ -128,6 +135,10 @@ func TestMain(m *testing.M) {
 }
 
 func TestAuditSeccomp(t *testing.T) {
+	if *deploySPOFlag == false {
+		t.Fatal("The Security Profiles Operator was not deployed. Please check that the deploy-spo flag is true")
+	}
+
```
Comment (blanquicet) [inline integration/integration_test.go:141]: I would prefer to let it fail if the SPO is not already installed in the system, don't you think?

File: integration/integration_test.go [hunk 2]
```diff
@@ -30,6 +30,8 @@ var (
 	image = flag.String("image", "", "gadget container image")
 
 	doNotDeploy = flag.Bool("no-deploy", false, "don't deploy Inspektor Gadget")
+
+	deploySPOFlag = flag.Bool("deploy-spo", true, "deploy the Security Profiles Operator (SPO)")
```
Comment (blanquicet) [inline integration/integration_test.go:34]: What about following the same approach for both and making it `no-deploy-spo`?

File: integration/integration_test.go [hunk 3]
```diff
@@ -89,16 +91,21 @@ func TestMain(m *testing.M) {
...
 		deployInspektorGadget,
-		deploySPO,
 		waitUntilInspektorGadgetPodsDeployed,
 		waitUntilInspektorGadgetPodsInitialized,
 	}
 
+	if *deploySPOFlag {
+		initCommands = append(initCommands, deploySPO)
+	}
+
```
Comment (blanquicet) [inline integration/integration_test.go:101]: What about decoupling `deploySPOFlag` from `doNotDeploy`? With this PR, the user cannot only deploy the SPO without deploying IG. I would suggest making them independent.

File: integration/integration_test.go [hunk 4]
```diff
@@ -30,6 +30,8 @@ var (
 	image = flag.String("image", "", "gadget container image")
 
 	doNotDeploy = flag.Bool("no-deploy", false, "don't deploy Inspektor Gadget")
```
Comment (blanquicet) [inline integration/integration_test.go:32]: nit: Maybe changing `doNotDeploy` to `doNotDeployIG` would make the code clearer now that we also have `deploySPOFlag`?

**Review by blanquicet — COMMENTED** [review]
I think this PR will make our lives easier! 🙂 ... Some comments but looks good overall.

File: integration/integration_test.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply integration/integration_test.go:141]: Yeah, I think it makes sense to let it fail.

File: integration/integration_test.go (hunk 2 above)
Comment (mauriciovasquezbernal) [reply integration/integration_test.go]: Right, updated to follow this approach.

File: integration/integration_test.go (hunk 3 above)
Comment (mauriciovasquezbernal) [reply integration/integration_test.go]: Makes sense, it was a bit complicated but I think it's fine now.

**Comment by mauriciovasquezbernal** [comment]
I updated the code following the suggestions. Please review with care, specially the "integration: Simplify logic around deferred calls" commit as it touches parts that have causes problems to us in the past.

File: integration/integration_test.go (hunk 3 above)
Comment (blanquicet) [reply integration/integration_test.go]: Yeah, it is clearer now. Thanks for managing this!

File: integration/integration_test.go [hunk 5]
```diff
@@ -64,19 +64,17 @@ func runCommands(cmds []*command, t *testing.T) {
 	}
```
Comment (eiffel-fl) [inline integration/integration_test.go:65]: Commit message:
`s/(guarante)/$1e/`
`s/(functions )all( called)/$1are$2/`

File: integration/integration_test.go [hunk 6]
```diff
@@ -29,7 +29,8 @@ var (
 	// image such as docker.io/kinvolk/gadget:latest
 	image = flag.String("image", "", "gadget container image")
 
-	doNotDeploy = flag.Bool("no-deploy", false, "don't deploy Inspektor Gadget")
+	doNotDeployIG  = flag.Bool("no-deploy-ig", false, "don't deploy Inspektor Gadget")
+	doNotDeploySPO = flag.Bool("no-deploy-spo", false, "don't deploy the Security Profiles Operator (SPO)")
```
Comment (eiffel-fl) [inline integration/integration_test.go:33]: Instead of `*DeploySPO`, can't we call this `*InstallSPO`?
So, we are not bothered regarding IG.

**Review by eiffel-fl — COMMENTED** [review]
I do not understand why we do not make `TestAuditSeccomp` conditionnal regarding the value of `doNotDeploySPO`.
It is a bit strange to know the test will fail but still run it if its requirements are not met...

**Comment by blanquicet** [comment]
> I do not understand why we do not make `TestAuditSeccomp` conditional regarding the value of `doNotDeploySPO`.
> It is a bit strange to know the test will fail but still run it if its requirements are not met...

I suggested this change because the fact that `-no-deploy-spo` is set does not mean the SPO is not installed in the cluster. Indeed, that's how contributors can make some quick local tests during the development. They install the SPO manually and then just execute `sudo -E go test -run TestAuditSeccomp -tags withebpf -test.v -integration -no-deploy-spo ./...` after each change to the `audit-seccomp` gadget. It will be indeed quicker given that installing/unsintalling the SPO takes a while.

**Comment by eiffel-fl** [comment]
> > I do not understand why we do not make `TestAuditSeccomp` conditional regarding the value of `doNotDeploySPO`.
> > It is a bit strange to know the test will fail but still run it if its requirements are not met...
> 
> I suggested this change because the fact that `-no-deploy-spo` is set does not mean the SPO is not installed in the cluster. Indeed, that's how contributors can make some quick local tests during the development. They install the SPO manually and then just execute `sudo -E go test -run TestAuditSeccomp -tags withebpf -test.v -integration -no-deploy-spo ./...` after each change to the `audit-seccomp` gadget. It will be indeed quicker given that installing/unsintalling the SPO takes a while.

I understand it is possible to install the SPO by yourself.
But I think we should add a check or at least an informative error message like "perhaps this is failing because you did not install the SPO?".

File: integration/integration_test.go (hunk 6 above)
Comment (mauriciovasquezbernal) [reply integration/integration_test.go:33]: Then why not "InstallIG"? I think both of them should have the prefix.

**Comment by mauriciovasquezbernal** [comment]
> But I think we should add a check or at least an informative error message like "perhaps this is failing because you did not install the SPO?".

It's deployed by default so I don't think it's a big problem. As Jose said I expected that somebody playing with the `no-deploy-spo` know what they are doing.

**Pull request merged**